src/include/qmpi.h

```diff
@@ -2556,6 +2561,54 @@ static inline int QMPI_Get_function(int calling_tool_id, enum QMPI_Functions_enu
     return mpi_errno;
 }
 
+static inline int QMPI_Get_calling_address(QMPI_Context context, void **address)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    void *array[BACKTRACE_SIZE];
+    char **traces;
+    int size, i;
+    int trace_entry = 1;        /* pick entry at 1 by default */
+    char *target = NULL;
+
+    /* Capture the calling function address */
+    size = backtrace(array, BACKTRACE_SIZE);
+    traces = backtrace_symbols(array, size);
+    if (traces != NULL) {
+        for (i = 0; i < size; i++) {
+            /* Skip PMPI_* calls */
+            char search[] = "PMPI_";
+            char *found = strstr(traces[i], search);
+            if (found != NULL) {
+                trace_entry += i;
+                break;
+            }
+        }
+        /* Extract the address from the backtrace */
+        if (size > trace_entry) {
+            const char *start_pattern = "[";
+            const char *end_pattern = "]";
+
+            char *start, *end;
+            if (start = strstr(traces[trace_entry], start_pattern)) {
+                start += strlen(start_pattern);
+                if (end = strstr(start, end_pattern)) {
+                    target = (char *) malloc(end - start + 1);
+                    memcpy(target, start, end - start);
+                    target[end - start] = '\0';
+                }
+            }
+        }
+    }
+    free(traces);
+    /* Store the calling address */
+    *address = calloc(0, strlen(target) + 1);
+    memcpy(*address, target, strlen(target) + 1);
+    free(target);
+
+    return mpi_errno;
+}
```

Comment:
Now that this function is no longer performance critical, I think it would be cleaner to put it in the `qmpi_register.c` file (I'll be splitting that file up soon to allow those functions to be intercepted) and avoid having to autogenerate it. Can you move it in there and just add the function declaration here instead?